crates/query-engine/translation/src/translation/query/root.rs

```diff
@@ -173,9 +221,126 @@ fn translate_rows(
+        // the limit clause is also applied here
+        let (order_by, order_by_joins) =
+            sorting::translate(env, state, &root_and_current_table, query.order_by.as_ref())?;
+
+        fields_select.order_by = order_by;
+        fields_select.joins.extend(order_by_joins);
+
+        // Add the limit.
+        fields_select.limit = sql::ast::Limit {
+            limit: query.limit,
+            offset: query.offset,
+        };
+    } else {
+        // if we aren't ordering across a relationship, we expect joins to be empty, and the order by and limit to be applied in the subquery
+        // however, we must repeat the order by clause, else ordering may not be guaranteed after joins
+        let (order_by, _order_by_joins) =
+            sorting::translate(env, state, &root_and_current_table, query.order_by.as_ref())?;
+        fields_select.order_by = order_by;
+    }
+
     Ok((returns_fields, fields_select))
 }
 
+/// Create a subquery for rows
+fn rows_subquery(
+    env: &Env,
+    state: &mut State,
+    make_from: &MakeFrom,
+    join_predicate: Option<&JoinPredicate<'_, '_>>,
+    query: &models::Query,
+) -> Result<(sql::ast::Select, TableSource), Error> {
+    let (current_table, from_clause) = make_reference_and_from_clause(env, state, make_from)?;
+
+    // the root table and the current table are the same at this point
+    let subquery_root_and_current_table = RootAndCurrentTables {
+        root_table: current_table.clone(),
+        current_table: current_table.clone(),
+    };
+
+    // we want to put the where clause, including any required joins, in a subquery that is applied before any joins used to navigate relationships
+    // this improves performance on cockroachdb
+    let mut subquery_select = sql::helpers::star_from_select(current_table.reference, from_clause);
+
+    // Apply a join predicate if we want one.
+    subquery_select.where_ = translate_where_with_join_predicate(
+        env,
+        state,
+        join_predicate,
+        query,
+        &subquery_root_and_current_table,
+    )?;
+
+    // unless there is an order by clause that traverses relationships, we can put the order by clause and limit in the subquery
+    if !order_by_crosses_relationships(query) {
+        // translate order_by
+        // we expect order_by_joins to be empty, because we're not traversing any relationship
+        let (order_by, _order_by_joins) = sorting::translate(
+            env,
+            state,
+            &subquery_root_and_current_table,
+            query.order_by.as_ref(),
+        )?;
+        subquery_select.order_by = order_by;
+        // Add the limit.
+        subquery_select.limit = sql::ast::Limit {
+            limit: query.limit,
+            offset: query.offset,
+        };
+    };
+
+    Ok((subquery_select, current_table.source))
+}
+
+fn translate_where_with_join_predicate(
```

Comment:
Part of the point of pulling this out was reusing it inside translate_query_part. That hasn't happened.

Reply:
Promise we'll get rid of that function entirely soon :)